File: gm-utils/src/main/java/src/commom/utils/doubles/DoubleIs.java
package src.commom.utils.doubles;

import src.commom.utils.object.Null;
import src.commom.utils.string.StringCompare;
import src.commom.utils.string.StringEmpty;
import src.commom.utils.string.StringExtraiCaracteresBaseadoEmString;
import src.commom.utils.string.StringOcorrencias;
import src.commom.utils.string.StringParse;

public class DoubleIs {
	
	private static final String NUMEROS = "0123456789.";

	public static boolean is(Object o) {

		if (Null.is(o)) {
			return false;
		}
		
		String s = StringParse.get(o);

		if (StringEmpty.is(s)) {
			return false;
		}

		if (!StringCompare.eq(s, StringExtraiCaracteresBaseadoEmString.exec(s, NUMEROS))) {
			return false;
		}
		
		return StringOcorrencias.get(s, ".") < 2;

	}

}

File: gm-utils/js/src/commom/utils/doubles/DoubleIs.js
import IntegerIs from '../integer/IntegerIs';
import Null from '../object/Null';
import StringCompare from '../string/StringCompare';
import StringContains from '../string/StringContains';
import StringEmpty from '../string/StringEmpty';
import StringExtraiCaracteres from '../string/StringExtraiCaracteres';
import StringExtraiNumeros from '../string/StringExtraiNumeros';
import StringParse from '../string/StringParse';
import StringSplit from '../string/StringSplit';

export default class DoubleIs {

	static is(o) {

		if (IntegerIs.is(o)) {
			return true;
		}

		if (Null.is(o)) {
			return false;
		}

		let s = StringParse.get(o);

		if (StringEmpty.is(s)) {
			return false;
		}

		let n = StringExtraiCaracteres.exec(s, DoubleIs.NUMEROS);

		if (!StringCompare.eq(s, n)) {
			return false;
		}

		let ints;
		let decs;

		if (StringContains.is(s, ".")) {

			if (s.startsWith(".")) {
				s = "0" + s;
			}

			let list = StringSplit.exec(s, ".");

			if (list.size() !== 2) {
				return false;
			}

			ints = list.get(0);
			decs = list.get(1);

		} else {
			ints = s;
			decs = "0";
		}

		return IntegerIs.is(ints) && IntegerIs.is(decs);

	}

}
DoubleIs.NUMEROS = StringExtraiNumeros.NUMEROS.concat2(".");
